python/pyarrow/dataset.py

```diff
@@ -121,27 +125,34 @@ def partitioning(field_names=None, flavor=None):
     """
     if flavor is None:
         # default flavor
-        if isinstance(field_names, pa.Schema):
-            return DirectoryPartitioning(field_names)
-        elif isinstance(field_names, list):
-            return DirectoryPartitioning.discover(field_names)
-        elif field_names is None:
-            raise ValueError(
-                "For the default directory flavor, need to specify "
-                "'field_names' as Schema or list of field names")
+        if schema is not None:
+            if field_names is not None:
+                raise ValueError(
+                    "Cannot specify both 'schema' and 'field_names'")
+            return DirectoryPartitioning(schema)
+        elif field_names is not None:
+            if isinstance(field_names, list):
+                return DirectoryPartitioning.discover(field_names)
+            else:
+                raise ValueError(
+                    "Expected list of field names, got {0}".format(
+                        type(field_names)))
         else:
             raise ValueError(
-                "Expected Schema or list of field names, got {0}".format(
-                    type(field_names)))
+                "For the default directory flavor, need to specify "
+                "a Schema or a list of field names")
     elif flavor == 'hive':
-        if isinstance(field_names, pa.Schema):
-            return HivePartitioning(field_names)
-        elif field_names is None:
-            return HivePartitioning.discover()
+        if field_names is not None:
+            raise ValueError("Cannot specify 'field_names' for flavor 'hive'")
+        elif schema is not None:
+            if isinstance(schema, pa.Schema):
+                return HivePartitioning(schema)
+            else:
+                raise ValueError(
```

Comment:
This is why I found useful the ability to restrict the HivePartitioning for certain field names. Perhaps I'm only interested in a subset of the partitioning fields, and it'd be easier to define them here.

@bkietz created an issue where we can discuss https://issues.apache.org/jira/browse/ARROW-7646